# Future Sales Prediction - Kaggle Competition

### Modeling File

https://www.kaggle.com/competitions/competitive-data-science-predict-future-sales/

#### File descriptions
- sales_train.csv - the training set. Daily historical data from January 2013 to October 2015.

- test.csv - the test set. You need to forecast the sales for these shops and products for November 2015.

- sample_submission.csv - a sample submission file in the correct format.

- items.csv - supplemental information about the items/products.

- item_categories.csv  - supplemental information about the items categories.

- shops.csv- supplemental information about the shops.

#### Data fields

- ID - an Id that represents a (Shop, Item) tuple within the test set

- shop_id - unique identifier of a shop

- item_id - unique identifier of a product

- item_category_id - unique identifier of item category

- item_cnt_day - number of products sold. You are predicting a monthly amount of this measure

- item_price - current price of an item

- date - date in format dd/mm/yyyy

- date_block_num - a consecutive month number, used for convenience. January 2013 is 0, February 2013 is 1,..., October 2015 is 33

- item_name - name of item

- shop_name - name of shop

- item_category_name - name of item category

### I. Import the dataset and data analysis libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
sales_train = pd.read_csv("data/sales_train.csv")
sales_train.head()

,date,date_block_num,shop_id,item_id,item_price,item_cnt_day
0,02.01.2013,0,59,22154,999.00,1.0
1,03.01.2013,0,25,2552,899.00,1.0
2,05.01.2013,0,25,2552,899.00,-1.0
3,06.01.2013,0,25,2554,1709.05,1.0
4,15.01.2013,0,25,2555,1099.00,1.0


In [3]:
test = pd.read_csv("data/test.csv")
test.head()

,ID,shop_id,item_id
0,0,5,5037
1,1,5,5320
2,2,5,5233
3,3,5,5232
4,4,5,5268


In [4]:
sample_submission = pd.read_csv("data/sample_submission.csv")
print(sample_submission.shape)
sample_submission.head()

(214200, 2)


,ID,item_cnt_month
0,0,0.5
1,1,0.5
2,2,0.5
3,3,0.5
4,4,0.5


In [5]:
items = pd.read_csv("data/items.csv")
items.head()

,item_name,item_id,item_category_id
0,! ВО ВЛАСТИ НАВАЖДЕНИЯ (ПЛАСТ.) D,0,40
1,!ABBYY FineReader 12 Professional Edition Full...,1,76
2,***В ЛУЧАХ СЛАВЫ (UNV) D,2,40
3,***ГОЛУБАЯ ВОЛНА (Univ) D,3,40
4,***КОРОБКА (СТЕКЛО) D,4,40


In [6]:
item_categories = pd.read_csv("data/item_categories.csv")
item_categories.head()

,item_category_name,item_category_id
0,PC - Гарнитуры/Наушники,0
1,Аксессуары - PS2,1
2,Аксессуары - PS3,2
3,Аксессуары - PS4,3
4,Аксессуары - PSP,4


In [7]:
shops = pd.read_csv("data/shops.csv")
shops.head()

,shop_name,shop_id
0,"!Якутск Орджоникидзе, 56 фран",0
1,"!Якутск ТЦ ""Центральный"" фран",1
2,"Адыгея ТЦ ""Мега""",2
3,"Балашиха ТРК ""Октябрь-Киномир""",3
4,"Волжский ТЦ ""Волга Молл""",4


### II. Initial data checks and manipulations

In [8]:
sales_train['date'] = pd.to_datetime(sales_train['date'], format = '%d.%m.%Y')
sales_train.head()

,date,date_block_num,shop_id,item_id,item_price,item_cnt_day
0,2013-01-02,0,59,22154,999.00,1.0
1,2013-01-03,0,25,2552,899.00,1.0
2,2013-01-05,0,25,2552,899.00,-1.0
3,2013-01-06,0,25,2554,1709.05,1.0
4,2013-01-15,0,25,2555,1099.00,1.0


In [9]:
# Split into month, day, and year
sales_train['year'] = sales_train['date'].dt.year
sales_train['month'] = sales_train['date'].dt.month
sales_train['day'] = sales_train['date'].dt.day
sales_train.head()

,date,date_block_num,shop_id,item_id,item_price,item_cnt_day,year,month,day
0,2013-01-02,0,59,22154,999.00,1.0,2013,1,2
1,2013-01-03,0,25,2552,899.00,1.0,2013,1,3
2,2013-01-05,0,25,2552,899.00,-1.0,2013,1,5
3,2013-01-06,0,25,2554,1709.05,1.0,2013,1,6
4,2013-01-15,0,25,2555,1099.00,1.0,2013,1,15


In [10]:
sales_train.isna().sum()

date              0
date_block_num    0
shop_id           0
item_id           0
item_price        0
item_cnt_day      0
year              0
month             0
day               0
dtype: int64

In [11]:
sales_train.nunique()

date               1034
date_block_num       34
shop_id              60
item_id           21807
item_price        19993
item_cnt_day        198
year                  3
month                12
day                  31
dtype: int64

In [12]:
# Describing datetime and quantitative variables
sales_train[['date', 'item_price', 'item_cnt_day']].describe()

,date,item_price,item_cnt_day
count,2935849,2.935849e+06,2.935849e+06
mean,2014-04-03 05:44:34.970681344,8.908532e+02,1.242641e+00
min,2013-01-01 00:00:00,-1.000000e+00,-2.200000e+01
25%,2013-08-01 00:00:00,2.490000e+02,1.000000e+00
50%,2014-03-04 00:00:00,3.990000e+02,1.000000e+00
75%,2014-12-05 00:00:00,9.990000e+02,1.000000e+00
max,2015-10-31 00:00:00,3.079800e+05,2.169000e+03
std,NaN,1.729800e+03,2.618834e+00


In [13]:
# Declaring categorical variables
sales_train['shop_id'] = pd.Categorical(sales_train['shop_id'])
sales_train['item_id'] = pd.Categorical(sales_train['item_id'])

In [14]:
# Convert negative to positive values in item_cnt_day
sales_train['item_cnt_day'] = sales_train['item_cnt_day'].abs()

In [15]:
# Create a "master table" for EDA in training data
sales_full = sales_train.merge(items, how = 'inner', on = 'item_id')\
                              .merge(item_categories, how = 'inner', on = 'item_category_id')\
                              .merge(shops, how = 'inner', on = 'shop_id')

sales_full.head()

,date,date_block_num,shop_id,item_id,item_price,item_cnt_day,year,month,day,item_name,item_category_id,item_category_name,shop_name
0,2013-01-02,0,59,22154,999.00,1.0,2013,1,2,ЯВЛЕНИЕ 2012 (BD),37,Кино - Blu-Ray,"Ярославль ТЦ ""Альтаир"""
1,2013-01-03,0,25,2552,899.00,1.0,2013,1,3,DEEP PURPLE The House Of Blue Light LP,58,Музыка - Винил,"Москва ТРК ""Атриум"""
2,2013-01-05,0,25,2552,899.00,1.0,2013,1,5,DEEP PURPLE The House Of Blue Light LP,58,Музыка - Винил,"Москва ТРК ""Атриум"""
3,2013-01-06,0,25,2554,1709.05,1.0,2013,1,6,DEEP PURPLE Who Do You Think We Are LP,58,Музыка - Винил,"Москва ТРК ""Атриум"""
4,2013-01-15,0,25,2555,1099.00,1.0,2013,1,15,DEEP PURPLE 30 Very Best Of 2CD (Фирм.),56,Музыка - CD фирменного производства,"Москва ТРК ""Атриум"""


#### Data checks:

1. Is there any item sold at more than one different price?

In [16]:
shops_and_items = sales_full[['shop_id', 'item_id', 'item_price']].copy()
shops_and_items.drop_duplicates(['shop_id', 'item_id'], inplace = True)

# Same shop, same item, different prices?
print(len(shops_and_items[shops_and_items.duplicated(subset = ['shop_id', 'item_id', 'item_price'])]))

0


2. Is there any group of items having different IDs but the same name? Do the same with shop name and item category name.

In [17]:
print(items.duplicated(['item_id']).sum()) # Number of duplicated item IDs
print(items.duplicated(['item_name']).sum()) # Number of duplicated item names

0
0


In [18]:
# With item categories
print(item_categories.duplicated(['item_category_id']).sum())
print(item_categories.duplicated(['item_category_name']).sum())

0
0


In [19]:
# With shops
print(shops.duplicated(['shop_id']).sum())
print(shops.duplicated(['shop_name']).sum())

0
0


3. Is there any day when the shop can't sell any item?

In [20]:
# Generate list of datetime indices - all days of sales in the training data
datetime_indices = sales_full[['date', 'date_block_num']].copy()
datetime_indices.drop_duplicates(inplace = True)
datetime_indices.sort_values('date', inplace = True)
datetime_indices.set_index('date', inplace = True)
print(datetime_indices.index)

DatetimeIndex(['2013-01-01', '2013-01-02', '2013-01-03', '2013-01-04',
               '2013-01-05', '2013-01-06', '2013-01-07', '2013-01-08',
               '2013-01-09', '2013-01-10',
               ...
               '2015-10-22', '2015-10-23', '2015-10-24', '2015-10-25',
               '2015-10-26', '2015-10-27', '2015-10-28', '2015-10-29',
               '2015-10-30', '2015-10-31'],
              dtype='datetime64[ns]', name='date', length=1034, freq=None)


In [21]:
# Generate a list of dates from the first and last day of sales in the training data
date_range = pd.date_range(start = sales_full['date'].min(), end = sales_full['date'].max())
print(date_range)

DatetimeIndex(['2013-01-01', '2013-01-02', '2013-01-03', '2013-01-04',
               '2013-01-05', '2013-01-06', '2013-01-07', '2013-01-08',
               '2013-01-09', '2013-01-10',
               ...
               '2015-10-22', '2015-10-23', '2015-10-24', '2015-10-25',
               '2015-10-26', '2015-10-27', '2015-10-28', '2015-10-29',
               '2015-10-30', '2015-10-31'],
              dtype='datetime64[ns]', length=1034, freq='D')


#### Fuzzy string matching - a part of data check:

Are there any pair of item names, shop names, or item category names close to each other? If this is possible, it might be the case of spelling or processing errors, or two categories which can be grouped to one. (Kaggle community said probably)

In [22]:
# https://typesense.org/learn/fuzzy-string-matching-python/
# https://pypi.org/project/fuzzywuzzy/
# https://towardsdatascience.com/a-fuzzy-string-matching-story-314bbecaa098/

from itertools import combinations, chain
from fuzzywuzzy import fuzz
from fuzzywuzzy import process

In [23]:
def fuzzy_string_matching_metrics(df, name_column):
    '''df: the DataFrame in question
    name_column: the column name that you want to check for fuzzy string matching (must be a string)'''

    name_list = df[name_column].tolist() # extract name pairs and convert them to a Python list
    pairwise_combinations = list(combinations(name_list, 2)) # generate all pairwise combinations of shop names to calculate similarity

    # https://stackoverflow.com/questions/7422453/python-change-type-of-whole-list
    # Change each element of the pairwise combinations into a list, then create a DataFrame for easier sorting
    pairwise_combinations = list(map(list, pairwise_combinations))
    pairwise_df = pd.DataFrame(pairwise_combinations, columns = ['first_string', 'second_string'])

    # Compute various similarity metrics
    pairwise_df['fuzz_ratio'] = pairwise_df.apply(lambda row: fuzz.ratio(row['first_string'], row['second_string']), axis = 1)
    pairwise_df['partial_ratio'] = pairwise_df.apply(lambda row: fuzz.partial_ratio(row['first_string'], row['second_string']), axis = 1)
    pairwise_df['token_sort_ratio'] = pairwise_df.apply(lambda row: fuzz.token_sort_ratio(row['first_string'], row['second_string']), axis = 1)
    pairwise_df['token_set_ratio'] = pairwise_df.apply(lambda row: fuzz.token_set_ratio(row['first_string'], row['second_string']), axis = 1)

    return pairwise_df

In [24]:
def sort_desc_with_threshold(df, sort_column, threshold):
    # Keep the necessary columns
    df_filtered = df[['first_string', 'second_string', sort_column]]

    # Filter for values greater or equal to a threshold 
    df_filtered = df_filtered[df_filtered[sort_column] >= threshold]

    # Sort values in descending order
    df_filtered.sort_values(sort_column, ascending = False, inplace = True)
    return df_filtered

In [25]:
def get_unique_values(*args):
    '''Pass individual list of lists as inputs. Returns a list of lists.'''

    # Combine the list of lists together and flatten it to be a 2D list
    whole_list = list(chain(args))
    flattened_list = [item for sublist in whole_list for item in sublist]

    # Convert each element of individual list to tuple format to later use the "set" data type (keeping unique values only)
    list_of_tuples = [tuple(x) for x in flattened_list]
    unique_val_set = set(list_of_tuples) # return a set object

    # Convert back into list of tuples and then list of lists
    unique_val_list = list(unique_val_set)
    unique_val_list = [list(x) for x in unique_val_list]

    return unique_val_list

In [26]:
def flag_similar_pairs(original_df, name_column, fuzz_ratio_threshold = 90, partial_ratio_threshold = 90,
                       token_sort_ratio_threshold = 90, token_set_ratio_threshold = 90):
    '''You can change the similarity thresholds by modifying the default arguments in this function.'''
    
    # Generate a DataFrame with fuzzy string matching metrics of all pairs of names in a particular DataFrame column
    df = fuzzy_string_matching_metrics(original_df, name_column)

    # Find similar pairs with high fuzz ratio, partial ratio, token sort ratio and token set ratio
    top_fuzz_ratios = sort_desc_with_threshold(df, 'fuzz_ratio', fuzz_ratio_threshold)[['first_string', 'second_string']].values.tolist()
    top_partial_ratios = sort_desc_with_threshold(df, 'partial_ratio', partial_ratio_threshold)[['first_string', 'second_string']].values.tolist()
    top_token_sort_ratios = sort_desc_with_threshold(df, 'token_sort_ratio', token_sort_ratio_threshold)[['first_string', 'second_string']].values.tolist()
    top_token_set_ratios = sort_desc_with_threshold(df, 'token_set_ratio', token_set_ratio_threshold)[['first_string', 'second_string']].values.tolist()

    # Get unique values that are high in at least one metric...
    unique_pairs_list = get_unique_values(top_fuzz_ratios, top_partial_ratios, top_token_sort_ratios, top_token_set_ratios)
    # ...then output a DataFrame for better readability
    result_df = df[df[['first_string', 'second_string']].apply(list, axis = 1).isin(unique_pairs_list)]
    return result_df

**For shop names:**

Because I have written so many functions, I decided to show the intermediate steps to arrive at finding similar pairs of names for this particular example.

In [27]:
shops.head()

,shop_name,shop_id
0,"!Якутск Орджоникидзе, 56 фран",0
1,"!Якутск ТЦ ""Центральный"" фран",1
2,"Адыгея ТЦ ""Мега""",2
3,"Балашиха ТРК ""Октябрь-Киномир""",3
4,"Волжский ТЦ ""Волга Молл""",4


In [28]:
# Example of a table of fuzzy string matching metrics
fuzzy_shop_names = fuzzy_string_matching_metrics(shops, 'shop_name')
fuzzy_shop_names.head()

,first_string,second_string,fuzz_ratio,partial_ratio,token_sort_ratio,token_set_ratio
0,"!Якутск Орджоникидзе, 56 фран","!Якутск ТЦ ""Центральный"" фран",52,52,53,59
1,"!Якутск Орджоникидзе, 56 фран","Адыгея ТЦ ""Мега""",22,33,29,29
2,"!Якутск Орджоникидзе, 56 фран","Балашиха ТРК ""Октябрь-Киномир""",20,21,33,33
3,"!Якутск Орджоникидзе, 56 фран","Волжский ТЦ ""Волга Молл""",23,25,33,33
4,"!Якутск Орджоникидзе, 56 фран","Вологда ТРЦ ""Мармелад""",24,28,34,34


In [29]:
# Example of keeping pairs with the highest fuzz ratio and sorting them in descending order
top_fuzz_ratio = sort_desc_with_threshold(fuzzy_shop_names, 'fuzz_ratio', 80)
top_fuzz_ratio

,first_string,second_string,fuzz_ratio
545,Жуковский ул. Чкалова 39м?,Жуковский ул. Чкалова 39м²,96
1104,"Москва ТК ""Буденовский"" (пав.А2)","Москва ТК ""Буденовский"" (пав.К7)",94
56,"!Якутск Орджоникидзе, 56 фран","Якутск Орджоникидзе, 56",88
115,"!Якутск ТЦ ""Центральный"" фран","Якутск ТЦ ""Центральный""",88
1560,"РостовНаДону ТРК ""Мегацентр Горизонт""","РостовНаДону ТРК ""Мегацентр Горизонт"" Островной",88
1335,"Москва ТЦ ""Перловский""","Москва ТЦ ""Семеновский""",84
1554,"Омск ТЦ ""Мега""","Химки ТЦ ""Мега""",83


In [30]:
# When the whole function is completed...
flag_similar_pairs(shops, 'shop_name')

,first_string,second_string,fuzz_ratio,partial_ratio,token_sort_ratio,token_set_ratio
56,"!Якутск Орджоникидзе, 56 фран","Якутск Орджоникидзе, 56",88,100,90,100
115,"!Якутск ТЦ ""Центральный"" фран","Якутск ТЦ ""Центральный""",88,100,89,100
545,Жуковский ул. Чкалова 39м?,Жуковский ул. Чкалова 39м²,96,96,100,100
1104,"Москва ТК ""Буденовский"" (пав.А2)","Москва ТК ""Буденовский"" (пав.К7)",94,94,89,94
1560,"РостовНаДону ТРК ""Мегацентр Горизонт""","РостовНаДону ТРК ""Мегацентр Горизонт"" Островной",88,100,88,100
1561,"РостовНаДону ТРК ""Мегацентр Горизонт""","РостовНаДону ТЦ ""Мега""",71,91,69,75
1580,"РостовНаДону ТРК ""Мегацентр Горизонт"" Островной","РостовНаДону ТЦ ""Мега""",61,91,58,75


There are some particularly interesting similar name combinations that have a high similarity (at least a similarity metric >= 90/100). I will use Google Translate to try to understand them.

- Index 56: ```!Якутск Орджоникидзе, 56 фран``` (!Yakutsk Ordzhonikidze, 56 fran) and ```Якутск Орджоникидзе, 56``` (Yakutsk Ordzhonikidze).

- Index 115: ```!Якутск ТЦ "Центральный" фран``` (!Yakutsk shopping center "Central" franc) and ```Якутск ТЦ "Центральный"``` (Yakutsk shopping center "Central").

- Index 545: ```Жуковский ул. Чкалова 39м?``` (Zhukovsky st. Chkalova 39m?) and ```Жуковский ул. Чкалова 39м²``` (Zhukovsky st. Chkalova 39m²).

- Index 1104: ```Москва ТК "Буденовский" (пав.А2)``` (Moscow TC "Budenovsky" (pavilion A2)) and ```Москва ТК "Буденовский" (пав.К7)``` (Moscow TC "Budenovsky" (pavement K7)).

- Index 1560, 1561 and 1580: ```РостовНаДону ТРК "Мегацентр Горизонт"``` (Rostov-on-Don TRC "Megacenter Gorizont"), ```РостовНаДону ТРК "Мегацентр Горизонт" Островной``` (Rostov-on-Don TRC "Megacenter Gorizont" Island) and ```РостовНаДону ТЦ "Мега"``` (Rostov-on-Don TC "Mega").

**Fuzzy string matching results:**

- ```фран``` (56, 115) does not have a clear meaning in Russian, so I think that this is a result of inaccurate processing or data entry.

- There are some unwanted special characters, such as ```!``` and ```?``` (56, 115, 545), that made spotting duplicated shop names impossible without fuzzy string matching.

- I have asked ChatGPT to save my time browsing through Russian sources while I don't understand anything about it, and it said that while ```РостовНаДону ТРК "Мегацентр Горизонт"``` and ```РостовНаДону ТРК "Мегацентр Горизонт" Островной``` (1560) refer to the same shop, ```РостовНаДону ТЦ "Мега"``` is another shopping center separated from these two names.

**For item category names:**

In [32]:
item_categories.head()

,item_category_name,item_category_id
0,PC - Гарнитуры/Наушники,0
1,Аксессуары - PS2,1
2,Аксессуары - PS3,2
3,Аксессуары - PS4,3
4,Аксессуары - PSP,4


In [33]:
flag_similar_pairs(item_categories, 'item_category_name')

,first_string,second_string,fuzz_ratio,partial_ratio,token_sort_ratio,token_set_ratio
83,Аксессуары - PS2,Аксессуары - PS3,94,94,93,93
84,Аксессуары - PS2,Аксессуары - PS4,94,94,93,93
85,Аксессуары - PS2,Аксессуары - PSP,94,94,93,93
86,Аксессуары - PS2,Аксессуары - PSVita,86,94,84,84
165,Аксессуары - PS3,Аксессуары - PS4,94,94,93,93
166,Аксессуары - PS3,Аксессуары - PSP,94,94,93,93
167,Аксессуары - PS3,Аксессуары - PSVita,86,94,84,84
246,Аксессуары - PS4,Аксессуары - PSP,94,94,93,93
247,Аксессуары - PS4,Аксессуары - PSVita,86,94,84,84
326,Аксессуары - PSP,Аксессуары - PSVita,86,94,84,84


**Decoding similar category names with Google Translate:**

| Russian      | English       | Russian   | English | Russian         | English       |
|--------------|---------------|-----------|---------|-----------------|---------------|
| Аксессуары   | Accessories   | Книги     | Books   | Игровые консоли | Game consoles |
| Подарки      | Gifts         | Игры      | Games   | Программы       | Programs      |
| Карты оплаты | Payment cards | Служебные | Utility | Кино            | Movies        |

**Special word combinations in Russian:**

- ```Книги - Аудиокниги (Цифра)```: Books - Audiobooks (Digital)

- ```Подарки - Настольные игры (компактные)```: Gifts - Board games (compact)

- ```Подарки - Сувениры (в навеску)```: Gifts - Souvenirs (as a gift)

- ```Программы - Для дома и офиса (Цифра)```: Programs - For home and office (Digital)

- ```Программы - Обучающие (Цифра)```: Programs - Educational (Digital)

- ```Служебные - Билеты```: Service - Tickets

**String similarity matching result:**

- Although being different types, ```Аксессуары``` (accessories), ```Игровые консоли``` (game consoles), and ```Игры``` (games) still have categories with the same names such as PS2, PS3, or PS4. For feature engineering, I can regroup each PS product into an item category.

- For the annotations (in the **Special word combinations** part):

    + ```Цифра``` (digital), ```компактные``` (compact), and ```Билеты``` (tickets) are additional information to distinguish between different item categories. I have decided to keep them as a separate category.

    + Because ```Подарки``` is already "gift" in Russian, the additional annotation ```в навеску``` (as a gift) is only extra information. I will merge ```Подарки - Сувениры``` and ```Подарки - Сувениры (в навеску)``` into a single category.

Python outputs ```MemoryError``` when I was working with item names (>20,000 different names), but I think that cleaning the shop names and item category names is enough.

### III. Feature engineering

Here, we consider ```date_block_num``` variable as number of months elapsed after January 2013, and this column can also be useful for splitting into different sets.

Aggregate existing features:

In [138]:
# Keep the necessary columns only - we are predicting by month
sales_features = sales_full[['year', 'month', 'date_block_num', 'shop_id', 'item_category_id', 'item_id', 'item_price', 'item_cnt_day']].copy()

# Because no item has more than 1 price, we can group item price by the median
sales_features_price = sales_features[['year', 'month', 'date_block_num', 'shop_id', 'item_category_id', 'item_id', 'item_price']]\
    .groupby(['year', 'month', 'date_block_num', 'shop_id', 'item_category_id', 'item_id'], as_index = False).median()

# Group item count by month by the sum
sales_features_itemcnt = sales_features[['year', 'month', 'date_block_num', 'shop_id', 'item_category_id', 'item_id', 'item_cnt_day']]\
    .groupby(['year', 'month', 'shop_id', 'item_category_id', 'item_id'], as_index = False).sum()

sales_features_itemcnt.rename(columns = {'item_cnt_day': 'item_cnt_month'}, inplace = True)

sales_features = sales_features_price.merge(sales_features_itemcnt, how = 'inner', on = ['year', 'month', 'date_block_num', 'shop_id', 'item_category_id', 'item_id'])
sales_features.sort_values(['year', 'month', 'date_block_num', 'shop_id', 'item_category_id', 'item_id', 'item_price', 'item_cnt_month'], inplace = True)
sales_features.head()

,year,month,date_block_num,shop_id,item_category_id,item_id,item_price,item_cnt_month
0,2013,1,0,0,2,5572,1322.0,10.0
1,2013,1,0,0,2,5573,560.0,1.0
2,2013,1,0,0,2,5575,806.0,4.0
3,2013,1,0,0,2,5576,2231.0,5.0
4,2013,1,0,0,2,5609,2381.0,1.0


Number of unique items in each shop/category for each month.

Median price of each shop/category/item.

In [139]:
# Number of unique items and median price in each shop

shops_and_items = sales_features[['year', 'month', 'shop_id', 'item_price', 'item_id']].copy()

shops_and_items = shops_and_items.groupby(['year', 'month', 'shop_id'], as_index = False)\
    .agg({'item_price': 'median', 'item_id': 'nunique'})
shops_and_items.rename(columns = {'item_price': 'med_price_shop', 'item_id': 'unique_items_in_shop'}, inplace = True)

shops_and_items.head()

,year,month,shop_id,med_price_shop,unique_items_in_shop
0,2013,1,0,196.0,2385
1,2013,1,1,190.0,1535
2,2013,1,2,399.0,728
3,2013,1,3,399.0,544
4,2013,1,4,299.0,1062


In [140]:
# Number of unique items and median price in each item category

cat_and_items = sales_features[['year', 'month', 'item_category_id', 'item_price', 'item_id']].copy()

cat_and_items = cat_and_items.groupby(['year', 'month', 'item_category_id'], as_index = False)\
    .agg({'item_price': 'median', 'item_id': 'nunique'})
cat_and_items.rename(columns = {'item_price': 'med_price_cat', 'item_id': 'unique_items_in_cat'}, inplace = True)

cat_and_items.head()

,year,month,item_category_id,med_price_cat,unique_items_in_cat
0,2013,1,0,148.0,1
1,2013,1,1,148.0,1
2,2013,1,2,2390.0,37
3,2013,1,3,499.0,2
4,2013,1,4,399.0,11


In [141]:
# Monthly median price of all items sold in the month in general

med_price_items = sales_features[['year', 'month', 'item_price']].copy()

med_price_items = med_price_items.groupby(['year', 'month'], as_index = False).median()
med_price_items.rename(columns = {'item_price': 'med_price_all_items'}, inplace = True)

med_price_items.head()

,year,month,med_price_all_items
0,2013,1,299.0
1,2013,2,299.0
2,2013,3,299.0
3,2013,4,299.0
4,2013,5,299.0


Create weekend and holiday features:

In [142]:
# Separate the sales days together
sales_days = sales_full[['date']].copy()
sales_days.drop_duplicates(inplace = True, ignore_index = True)
sales_days.sort_values('date', inplace = True, ignore_index = True)

# Create month, year, and weekend indicator variables
sales_days['year'] = sales_days['date'].dt.year
sales_days['month'] = sales_days['date'].dt.month
sales_days['weekend_indicator'] = sales_days['date'].case_when(
    [(sales_days['date'].dt.day_of_week.isin([5, 6]), 1),
     (~sales_days['date'].dt.day_of_week.isin([5, 6]), 0)]
)

sales_days.head(7)

,date,year,month,weekend_indicator
0,2013-01-01,2013,1,0
1,2013-01-02,2013,1,0
2,2013-01-03,2013,1,0
3,2013-01-04,2013,1,0
4,2013-01-05,2013,1,1
5,2013-01-06,2013,1,1
6,2013-01-07,2013,1,0


In [143]:
# See Analysis File for more details
# List of holidays in Russia: https://www.timeanddate.com/holidays/russia/2013 (same as for 2014 and 2015)
holiday_list_2013 = ['2013-01-01', '2013-01-02', '2013-01-03', '2013-01-04', '2013-01-05', '2013-01-06', '2013-01-07', '2013-01-08',
                     '2013-02-23', '2013-03-08', '2013-05-01', '2013-05-02', '2013-05-03', '2013-05-09', '2013-05-10',
                     '2013-06-12', '2013-11-04']

holiday_list_2014 = ['2014-01-01', '2014-01-02', '2014-01-03', '2014-01-06', '2014-01-07', '2014-01-08',
                     '2014-02-22', '2014-02-23', '2014-03-08', '2014-03-09', '2014-03-10', '2014-05-01', '2014-05-02',
                     '2014-05-03', '2014-05-09', '2014-05-10', '2014-05-11', '2014-06-12', '2014-06-13', '2014-06-14', '2014-06-15',
                     '2014-11-01', '2014-11-02', '2014-11-03', '2014-11-04']

holiday_list_2015 = ['2015-01-01', '2015-01-02', '2015-01-03', '2015-01-04', '2015-01-05', '2015-01-06', '2015-01-07',
                     '2015-01-08', '2015-01-09', '2015-02-23', '2015-03-08', '2015-03-09', '2015-05-01', '2015-05-04', '2015-05-09',
                     '2015-05-11', '2015-06-12'] # only count until October 2015

# Combine to have a list of holidays in training data
holidays = holiday_list_2013 + holiday_list_2014 + holiday_list_2015
holidays = [pd.Timestamp(x) for x in holidays] # Convert the whole list to timestamp format

# Create holiday indicator variable
sales_days['holiday_indicator'] = sales_days['date'].case_when(
    [(sales_days['date'].isin(holidays), 1),
     (~sales_days['date'].isin(holidays), 0)]
)

sales_days.head()

,date,year,month,weekend_indicator,holiday_indicator
0,2013-01-01,2013,1,0,1
1,2013-01-02,2013,1,0,1
2,2013-01-03,2013,1,0,1
3,2013-01-04,2013,1,0,1
4,2013-01-05,2013,1,1,1


In [144]:
# Count the number of days as weekends and holidays in a (month, year) pair
sales_days.drop('date', axis = 1, inplace = True)
sales_month = sales_days.groupby(['year', 'month'], as_index = False).sum()
sales_month.rename(columns = {'weekend_indicator': 'weekends_in_month', 'holiday_indicator': 'holidays_in_month'}, inplace = True)
sales_month.head(7)

,year,month,weekends_in_month,holidays_in_month
0,2013,1,8,8
1,2013,2,8,1
2,2013,3,10,1
3,2013,4,8,0
4,2013,5,8,5
5,2013,6,10,1
6,2013,7,8,0


Create seasonal features:

In [145]:
sales_month['season'] = sales_month['month'].case_when(
    [(sales_month['month'].isin([3, 4, 5]), 'spring'),
     (sales_month['month'].isin([6, 7, 8]), 'summer'),
     (sales_month['month'].isin([9, 10, 11]), 'fall'),
     (sales_month['month'].isin([12, 1, 2]), 'winter')]
)

# Because winter is very cold in Russia, I will encode so that winter goes first
sales_month['season'] = pd.Categorical(values = sales_month['season'], ordered = True,
                                       categories = ['winter', 'spring', 'summer', 'fall'])

# Use one-hot encoding to turn the seasons into features
# Drop the first column (winter) to avoid multicollinearity
onehot_season = pd.get_dummies(sales_month['season'], drop_first = True)

# Convert data type of all the one-hot encoded values: True/False to 0/1
onehot_season = onehot_season.map(lambda x: int(x))

# Merge the one-hot encoding into the sales_month feature DataFrame
sales_month = pd.concat([sales_month, onehot_season], axis = 1)

# Drop the categorical 'season' column now that we have 3 one-hot encoded season columns
sales_month.drop('season', axis = 1, inplace = True)

sales_month.head(7)

,year,month,weekends_in_month,holidays_in_month,spring,summer,fall
0,2013,1,8,8,0,0,0
1,2013,2,8,1,0,0,0
2,2013,3,10,1,1,0,0
3,2013,4,8,0,1,0,0
4,2013,5,8,5,1,0,0
5,2013,6,10,1,0,1,0
6,2013,7,8,0,0,1,0


Now combine the features together:

In [146]:
monthly_features = shops_and_items\
    .merge(cat_and_items, how = 'inner', on = ['year', 'month'])\
    .merge(med_price_items, how = 'inner', on = ['year', 'month'])\
    .merge(sales_month, how = 'inner', on = ['year', 'month'])

monthly_features.head(7)

,year,month,shop_id,med_price_shop,unique_items_in_shop,item_category_id,med_price_cat,unique_items_in_cat,med_price_all_items,weekends_in_month,holidays_in_month,spring,summer,fall
0,2013,1,0,196.0,2385,0,148.00,1,299.0,8,8,0,0,0
1,2013,1,0,196.0,2385,1,148.00,1,299.0,8,8,0,0,0
2,2013,1,0,196.0,2385,2,2390.00,37,299.0,8,8,0,0,0
3,2013,1,0,196.0,2385,3,499.00,2,299.0,8,8,0,0,0
4,2013,1,0,196.0,2385,4,399.00,11,299.0,8,8,0,0,0
5,2013,1,0,196.0,2385,5,699.00,27,299.0,8,8,0,0,0
6,2013,1,0,196.0,2385,6,1493.75,19,299.0,8,8,0,0,0


In [147]:
final_features = sales_features.merge(monthly_features, how = 'left', on = ['year', 'month', 'shop_id', 'item_category_id'])

final_features.head(7)

,year,month,date_block_num,shop_id,item_category_id,item_id,item_price,item_cnt_month,med_price_shop,unique_items_in_shop,med_price_cat,unique_items_in_cat,med_price_all_items,weekends_in_month,holidays_in_month,spring,summer,fall
0,2013,1,0,0,2,5572,1322.0,10.0,196.0,2385,2390.0,37,299.0,8,8,0,0,0
1,2013,1,0,0,2,5573,560.0,1.0,196.0,2385,2390.0,37,299.0,8,8,0,0,0
2,2013,1,0,0,2,5575,806.0,4.0,196.0,2385,2390.0,37,299.0,8,8,0,0,0
3,2013,1,0,0,2,5576,2231.0,5.0,196.0,2385,2390.0,37,299.0,8,8,0,0,0
4,2013,1,0,0,2,5609,2381.0,1.0,196.0,2385,2390.0,37,299.0,8,8,0,0,0
5,2013,1,0,0,2,5612,3623.0,1.0,196.0,2385,2390.0,37,299.0,8,8,0,0,0
6,2013,1,0,0,2,5623,294.0,1.0,196.0,2385,2390.0,37,299.0,8,8,0,0,0


The last preprocessing before I start to train my models:

In [148]:
# Encode month using trigonometric encoding for cyclical features
final_features['month_sin'] = np.sin(final_features['month'] / 12 * 2 * np.pi)
final_features['month_cos'] = np.cos(final_features['month'] / 12 * 2 * np.pi)

# Drop the month variable now that it isn't necessary
final_features.drop('month', axis = 1, inplace = True)

final_features.head(7)

,year,date_block_num,shop_id,item_category_id,item_id,item_price,item_cnt_month,med_price_shop,unique_items_in_shop,med_price_cat,unique_items_in_cat,med_price_all_items,weekends_in_month,holidays_in_month,spring,summer,fall,month_sin,month_cos
0,2013,0,0,2,5572,1322.0,10.0,196.0,2385,2390.0,37,299.0,8,8,0,0,0,0.5,0.866025
1,2013,0,0,2,5573,560.0,1.0,196.0,2385,2390.0,37,299.0,8,8,0,0,0,0.5,0.866025
2,2013,0,0,2,5575,806.0,4.0,196.0,2385,2390.0,37,299.0,8,8,0,0,0,0.5,0.866025
3,2013,0,0,2,5576,2231.0,5.0,196.0,2385,2390.0,37,299.0,8,8,0,0,0,0.5,0.866025
4,2013,0,0,2,5609,2381.0,1.0,196.0,2385,2390.0,37,299.0,8,8,0,0,0,0.5,0.866025
5,2013,0,0,2,5612,3623.0,1.0,196.0,2385,2390.0,37,299.0,8,8,0,0,0,0.5,0.866025
6,2013,0,0,2,5623,294.0,1.0,196.0,2385,2390.0,37,299.0,8,8,0,0,0,0.5,0.866025


- Khong nen cho year, date_block_num, id... vào dữ liệu
- Làm thêm feature: trung bình item_cnt_month 1 tháng
- Cột tổng doanh thu, trung bình doanh thu,...

#### Explanation of new feature names:

- ```med_price_shop```, ```unique_items_in_shop```: Median price of an item and number of unique items sold in a particular shop in the month

- ```med_price_cat```, ```unique_items_in_cat```: Median price of an item and number of unique items belonging to a particular category sold in the month

- ```med_price_all_items```: Median price of all items sold in the month

- ```weekends_in_month```: Number of days of weekend (Saturday or Sunday) in the month

- ```holidays_in_month```: Number of days of public holiday in the month

- ```spring```, ```summer```, ```fall```: Dummy variables denoting the season that an item was sold. If ```spring = summer = fall = 0```, then this item was sold in winter

- ```month_sin```, ```month_cos```: Month in a year, encoded using trigonometric encoding to reflect its cyclical nature throughout a year.

### IV. Declare training and validation sets

We are using 5-fold time series split cross-validation. However, because there are more than 1 observation in a single time frame (panel data), I have to split the folds manually.

Splitting plan (using date_block_num): There are 34 months in the training data (indexed from 0 to 33). I will split the folds so that for the smallest fold, there is at least sufficient data for a year to capture the cyclical characteristics of time series.

| Fold | Fold's training data          | Fold's validation data   |
|------|-------------------------------|--------------------------|
| 1    | ```date_block_num```: 0 -> 11 | ```date_block_num```: 12 |
| 2    | ```date_block_num```: 0 -> 17 | ```date_block_num```: 18 |
| 3    | ```date_block_num```: 0 -> 23 | ```date_block_num```: 24 |
| 4    | ```date_block_num```: 0 -> 29 | ```date_block_num```: 30 |
| 5    | ```date_block_num```: 0 -> 32 | ```date_block_num```: 33 |

In [149]:
# Split into the corresponding folds

# Fold 1
fold_1_train = final_features[final_features['date_block_num'].isin(range(12))]
fold_1_test = final_features[final_features['date_block_num'] == 12]

# Fold 2
fold_2_train = final_features[final_features['date_block_num'].isin(range(18))]
fold_2_test = final_features[final_features['date_block_num'] == 18]

# Fold 3
fold_3_train = final_features[final_features['date_block_num'].isin(range(24))]
fold_3_test = final_features[final_features['date_block_num'] == 24]

# Fold 4
fold_4_train = final_features[final_features['date_block_num'].isin(range(30))]
fold_4_test = final_features[final_features['date_block_num'] == 30]

# Fold 5
fold_5_train = final_features[final_features['date_block_num'].isin(range(33))]
fold_5_test = final_features[final_features['date_block_num'] == 33]

In [150]:
# Number of data points in each fold
print(f'''Fold 1: {fold_1_train.shape[0]} items in training set, {fold_1_test.shape[0]} items in validation set.
      Validation / Training = {fold_1_test.shape[0] / fold_1_train.shape[0]}''')
print('-------------------------------------------------------------')
print(f'''Fold 2: {fold_2_train.shape[0]} items in training set, {fold_2_test.shape[0]} items in validation set.
      Validation / Training = {fold_2_test.shape[0] / fold_2_train.shape[0]}''')
print('-------------------------------------------------------------')
print(f'''Fold 3: {fold_3_train.shape[0]} items in training set, {fold_3_test.shape[0]} items in validation set.
      Validation / Training = {fold_3_test.shape[0] / fold_3_train.shape[0]}''')
print('-------------------------------------------------------------')
print(f'''Fold 4: {fold_4_train.shape[0]} items in training set, {fold_4_test.shape[0]} items in validation set.
      Validation / Training = {fold_4_test.shape[0] / fold_4_train.shape[0]}''')
print('-------------------------------------------------------------')
print(f'''Fold 5: {fold_5_train.shape[0]} items in training set, {fold_5_test.shape[0]} items in validation set.
      Validation / Training = {fold_5_test.shape[0] / fold_5_train.shape[0]}''')

Fold 1: 482291 items in training set, 35056 items in validation set.
      Validation / Training = 0.07268640716911574
-------------------------------------------------------------
Fold 2: 674998 items in training set, 31685 items in validation set.
      Validation / Training = 0.04694087982482911
-------------------------------------------------------------
Fold 3: 857397 items in training set, 29839 items in validation set.
      Validation / Training = 0.03480184791875875
-------------------------------------------------------------
Fold 4: 1008819 items in training set, 23354 items in validation set.
      Validation / Training = 0.023149841547393538
-------------------------------------------------------------
Fold 5: 1075730 items in training set, 21820 items in validation set.
      Validation / Training = 0.020283900235189126


In [151]:
# Split into features and targets

# Fold 1
X_train_1 = fold_1_train.drop('item_cnt_month', axis = 1).values
y_train_1 = fold_1_train[['item_cnt_month']].values.ravel()
X_test_1 = fold_1_test.drop('item_cnt_month', axis = 1).values
y_test_1 = fold_1_test[['item_cnt_month']].values.ravel()

# Fold 2
X_train_2 = fold_2_train.drop('item_cnt_month', axis = 1).values
y_train_2 = fold_2_train[['item_cnt_month']].values.ravel()
X_test_2 = fold_2_test.drop('item_cnt_month', axis = 1).values
y_test_2 = fold_2_test[['item_cnt_month']].values.ravel()

# Fold 3
X_train_3 = fold_3_train.drop('item_cnt_month', axis = 1).values
y_train_3 = fold_3_train[['item_cnt_month']].values.ravel()
X_test_3 = fold_3_test.drop('item_cnt_month', axis = 1).values
y_test_3 = fold_3_test[['item_cnt_month']].values.ravel()

# Fold 4
X_train_4 = fold_4_train.drop('item_cnt_month', axis = 1).values
y_train_4 = fold_4_train[['item_cnt_month']].values.ravel()
X_test_4 = fold_4_test.drop('item_cnt_month', axis = 1).values
y_test_4 = fold_4_test[['item_cnt_month']].values.ravel()

# Fold 5
X_train_5 = fold_5_train.drop('item_cnt_month', axis = 1).values
y_train_5 = fold_5_train[['item_cnt_month']].values.ravel()
X_test_5 = fold_5_test.drop('item_cnt_month', axis = 1).values
y_test_5 = fold_5_test[['item_cnt_month']].values.ravel()

In [152]:
# Check the shape of each X and y in training and test sets
print(f'''Fold 1: X_train: {X_train_1.shape}, y_train: {y_train_1.shape}, X_test: {X_test_1.shape}, y_test: {y_test_1.shape}''')
print(f'''Fold 2: X_train: {X_train_2.shape}, y_train: {y_train_2.shape}, X_test: {X_test_2.shape}, y_test: {y_test_2.shape}''')
print(f'''Fold 3: X_train: {X_train_3.shape}, y_train: {y_train_3.shape}, X_test: {X_test_3.shape}, y_test: {y_test_3.shape}''')
print(f'''Fold 4: X_train: {X_train_4.shape}, y_train: {y_train_4.shape}, X_test: {X_test_4.shape}, y_test: {y_test_4.shape}''')
print(f'''Fold 5: X_train: {X_train_5.shape}, y_train: {y_train_5.shape}, X_test: {X_test_5.shape}, y_test: {y_test_5.shape}''')

Fold 1: X_train: (482291, 18), y_train: (482291,), X_test: (35056, 18), y_test: (35056,)
Fold 2: X_train: (674998, 18), y_train: (674998,), X_test: (31685, 18), y_test: (31685,)
Fold 3: X_train: (857397, 18), y_train: (857397,), X_test: (29839, 18), y_test: (29839,)
Fold 4: X_train: (1008819, 18), y_train: (1008819,), X_test: (23354, 18), y_test: (23354,)
Fold 5: X_train: (1075730, 18), y_train: (1075730,), X_test: (21820, 18), y_test: (21820,)


In [153]:
# Example on fold 1
fold_1_train.tail()

,year,date_block_num,shop_id,item_category_id,item_id,item_price,item_cnt_month,med_price_shop,unique_items_in_shop,med_price_cat,unique_items_in_cat,med_price_all_items,weekends_in_month,holidays_in_month,spring,summer,fall,month_sin,month_cos
482286,2013,11,59,75,3160,1590.0,1.0,299.0,671,1290.0,96,299.0,9,0,0,0,0,-2.449294e-16,1.0
482287,2013,11,59,75,4186,1491.0,1.0,299.0,671,1290.0,96,299.0,9,0,0,0,0,-2.449294e-16,1.0
482288,2013,11,59,75,4688,1499.0,1.0,299.0,671,1290.0,96,299.0,9,0,0,0,0,-2.449294e-16,1.0
482289,2013,11,59,75,12727,1490.0,1.0,299.0,671,1290.0,96,299.0,9,0,0,0,0,-2.449294e-16,1.0
482290,2013,11,59,83,22091,109.0,1.0,299.0,671,79.0,6,299.0,9,0,0,0,0,-2.449294e-16,1.0


In [154]:
# Set seed for reproducibility
my_seed = 275225

In [155]:
# What features can I apply scaling on (with MinMaxScaler)?

features_to_remove = ['shop_id', 'item_category_id', 'item_id', 'month_sin', 'month_cos', 'spring', 'summer', 'fall']
features_list = final_features.columns.tolist().copy() # start with all features that I have included

# Need to use an if statement so that this code can be rerun many times (this includes inplace removal of list elements)
for feature in features_to_remove:
    if feature in features_list:
        features_list.remove(feature)

print(features_list)

['year', 'date_block_num', 'item_price', 'item_cnt_month', 'med_price_shop', 'unique_items_in_shop', 'med_price_cat', 'unique_items_in_cat', 'med_price_all_items', 'weekends_in_month', 'holidays_in_month']


In [156]:
# https://stackoverflow.com/questions/22934609/get-indexes-of-multiple-pandas-columns-by-names
# Get index of the columns to pass to ColumnTransformer without error
# https://stackoverflow.com/questions/71715754/valueerror-specifying-the-columns-using-strings-is-only-supported-for-pandas-da

categorical_col_indices = [final_features.columns.get_loc(col) for col in ['shop_id', 'item_category_id', 'item_id']]
print(f'Indices of categorical columns: {categorical_col_indices}')

minmax_indices = [final_features.columns.get_loc(col) for col in features_list]
print(f'Indices of columns to apply MinMaxScaler on: {minmax_indices}')

Indices of categorical columns: [2, 3, 4]
Indices of columns to apply MinMaxScaler on: [0, 1, 5, 6, 7, 8, 9, 10, 11, 12, 13]


In [157]:
from sklearn.preprocessing import TargetEncoder, MinMaxScaler
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

# Encode categorical variables (shop_id, item_category_id, item_id) with TargetEncoder
# Encode all other variables (except for "month" and dummy variables) with MinMaxScaler, default range [0, 1]

target_enc = TargetEncoder(target_type = 'continuous', shuffle = True, random_state = my_seed)

transformer = ColumnTransformer(
    transformers = [('target_enc', target_enc, categorical_col_indices),
                    ('minmax_scaler', MinMaxScaler(), minmax_indices)],
    remainder = 'passthrough'
)

### V. Implementing regression models and assessing feature importance

In [158]:
from sklearn.metrics import root_mean_squared_error
import statistics
import time # I want to measure the efficiency of different machine learning models

# Custom function for training in many folds and many different models

folds = [[X_train_1, X_test_1, y_train_1, y_test_1], [X_train_2, X_test_2, y_train_2, y_test_2],
         [X_train_3, X_test_3, y_train_3, y_test_3], [X_train_4, X_test_4, y_train_4, y_test_4],
         [X_train_5, X_test_5, y_train_5, y_test_5]]

def kfold_regression(model_name, model, transformer_x, folds, output = 'df'):
    # Choices of output:
    # (1) 'df' (default, output a neatly formatted DataFrame of the results)
    # (2) 'score' (output only the mean RMSE)
    # (3) 'model' (output the model)

    rmse_folds = []

    pipe = Pipeline([('transformer', transformer_x), ('regression', model)])

    # I want to measure the time that Python fits the data and output the scores for all 5 folds
    start_time = time.perf_counter()

    # Each element in "folds" list should have the form [X_train, X_test, y_train, y_test]
    for fold in folds:
        pipe.fit(fold[0], fold[2])
        y_pred = pipe.predict(fold[1])
        rmse_folds.append(root_mean_squared_error(fold[3], y_pred))
    
    end_time = time.perf_counter()

    if output == 'df':
        df = pd.DataFrame({'model_name': [model_name], 'mean_rmse': [statistics.mean(rmse_folds)],
                       'std_rmse': [statistics.stdev(rmse_folds)], 'execution_time': [end_time - start_time]})
        return df
    elif output == 'score':
        return statistics.mean(rmse_folds)
    elif output == 'model':
        return pipe['regression']
    else:
        raise ValueError('The "output" parameter must be \'df\', \'score\', or \'model\'.')

In [159]:
# Import the necessary models
from sklearn.linear_model import LinearRegression, Lasso, Ridge, ElasticNet
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, AdaBoostRegressor, GradientBoostingRegressor, HistGradientBoostingRegressor
from xgboost import XGBRegressor
from catboost import CatBoostRegressor
from lightgbm import LGBMRegressor

In [160]:
import re
import warnings
warnings.filterwarnings(action = 'ignore', message = r'^X does not have valid feature names')

In [ ]:
model_list = [('Linear Regression', LinearRegression(n_jobs = -1), transformer, folds),
               ('Lasso', Lasso(random_state = my_seed), transformer, folds),
               ('Ridge', Ridge(random_state = my_seed), transformer, folds),
               ('Elastic Net', ElasticNet(random_state = my_seed), transformer, folds)]

overall_results = pd.concat([kfold_regression(model_name, model, transformer, folds) for model_name, model, transformer, folds in model_list], axis = 0, ignore_index = True)
overall_results.sort_values(['mean_rmse', 'std_rmse', 'execution_time'], inplace = True, ignore_index = True)
overall_results

,model_name,mean_rmse,std_rmse,execution_time
0,Elastic Net,1.995931,2.520029,12.761016
1,Lasso,1.995931,2.520029,38.955910
2,Ridge,2.093282,2.446418,19.947494
3,Linear Regression,2.093299,2.446407,36.233729


In [ ]:
# WARNING: THIS CODE BLOCK WILL TAKE A LONG TIME TO RUN (est. > 40 mins)

model_list_2 = [('Decision Tree', DecisionTreeRegressor(random_state = my_seed), transformer, folds),
              ('Random Forest', RandomForestRegressor(n_jobs = -1, random_state = my_seed), transformer, folds),
              ('AdaBoost', AdaBoostRegressor(random_state = my_seed), transformer, folds),
              ('Gradient Boosting', GradientBoostingRegressor(random_state = my_seed), transformer, folds),
              ('Hist-based Gradient Boosting', HistGradientBoostingRegressor(random_state = my_seed), transformer, folds),
              ('XGBoost', XGBRegressor(n_jobs = -1, random_state = my_seed), transformer, folds),
              ('CatBoost', CatBoostRegressor(random_state = my_seed, verbose = False), transformer, folds),
              ('LightGBM', LGBMRegressor(n_jobs = -1, random_state = my_seed, verbose = -1), transformer, folds)]

overall_results = pd.concat([kfold_regression(model_name, model, transformer, folds) for model_name, model, transformer, folds in model_list_2], axis = 0, ignore_index = True)
overall_results.sort_values(['mean_rmse', 'std_rmse', 'execution_time'], inplace = True, ignore_index = True)
overall_results

,model_name,mean_rmse,std_rmse,execution_time
0,Gradient Boosting,1.314370,1.584073,995.316409
1,Random Forest,1.414142,1.761584,529.898067
2,LightGBM,1.469164,1.988627,16.464381
3,CatBoost,1.647621,1.750774,217.158482
4,Decision Tree,1.683901,1.632204,182.598874
5,AdaBoost,1.727030,1.885637,432.293311
6,Hist-based Gradient Boosting,1.902756,2.348932,20.536378
7,XGBoost,2.738961,2.255959,17.657261


### VI. Assessing feature importance

In [108]:
def coef_or_feature_importances(model, folds, output = 'df'):
    # output takes 2 possible values: 'df' (pandas DataFrame) or 'np' (NumPy array)

    importances_folds = []

    pipe = Pipeline([('transformer', transformer), ('regression', model)])

    # Each element in "folds" list should have the form [X_train, X_test, y_train, y_test]
    for fold in folds:
        pipe.fit(fold[0], fold[2])

        try:
            importances_folds.append(pipe['regression'].coef_)

        except AttributeError: # for tree-based models which do not have coefficients attribute
            importances_folds.append(pipe['regression'].feature_importances_)
    
    if (output == 'df') or (output == 'np'):
        importances_folds = np.array(importances_folds)
        if output == 'df':
            importances_folds = pd.DataFrame(importances_folds)
            # Transpose the DataFrame so that rows = features, columns = folds
            importances_folds = importances_folds.transpose()
            # Give the features a name
            importances_folds.index = fold_1_train.drop('item_cnt_month', axis = 1).columns
            importances_folds['mean_coef_or_importance'] = importances_folds.mean(axis = 1)
            importances_folds.sort_values('mean_coef_or_importance', ascending = False, inplace = True)

    return importances_folds

#### Linear-based models:

In [ ]:
# Mean absolute coefficient across 5 folds of Linear Regression (top 7 only)
linreg_coef = coef_or_feature_importances(LinearRegression(n_jobs = -1), folds)
linreg_coef['abs_mean_coef'] = linreg_coef['mean_coef_or_importance'].abs()
linreg_coef.sort_values('abs_mean_coef', ascending = False, inplace = True)
linreg_coef.iloc[:7]

,0,1,2,3,4,mean_coef_or_importance,abs_mean_coef
med_price_shop,0.289062,0.364065,0.616721,4.044620,3.885479,1.839989,1.839989
shop_id,1.212450,1.158344,1.110352,1.037005,1.037697,1.111170,1.111170
item_price,1.224198,1.587866,1.238001,0.626277,0.536267,1.042522,1.042522
unique_items_in_shop,0.927807,0.827209,0.808218,0.815631,0.803608,0.836494,0.836494
holidays_in_month,0.448222,0.524611,0.357545,0.325824,0.315205,0.394281,0.394281
item_id,-0.282016,-0.168054,-0.468775,-0.450819,-0.563629,-0.386659,0.386659
date_block_num,0.209291,0.272792,0.326647,0.375655,0.389165,0.314710,0.314710


In [118]:
# Mean absolute coefficient across 5 folds of Ridge Regression (top 7 only)
ridge_coef = coef_or_feature_importances(Ridge(random_state = my_seed), folds)
ridge_coef['abs_mean_coef'] = ridge_coef['mean_coef_or_importance'].abs()
ridge_coef.sort_values('abs_mean_coef', ascending = False, inplace = True)
ridge_coef.iloc[:7]

,0,1,2,3,4,mean_coef_or_importance,abs_mean_coef
med_price_shop,0.289467,0.364502,0.616854,3.993236,3.840205,1.820853,1.820853
shop_id,1.212453,1.158330,1.110327,1.037005,1.037699,1.111163,1.111163
unique_items_in_shop,0.927765,0.827205,0.808174,0.814444,0.802537,0.836025,0.836025
item_price,0.887063,1.224976,1.007409,0.533210,0.462489,0.823030,0.823030
holidays_in_month,0.448198,0.524570,0.357554,0.326010,0.315362,0.394339,0.394339
item_id,-0.282019,-0.168098,-0.468668,-0.450213,-0.563047,-0.386409,0.386409
date_block_num,0.209247,0.272762,0.326631,0.375755,0.389255,0.314730,0.314730


I couldn't calculate feature importance (through coefficient) with Lasso and ElasticNet regression because the models zeroed out all coefficients, making comparison very difficult.

#### Tree-based models:

In [115]:
# Mean feature importances across 5 folds of LightGBM (top 7 only)
lgbm_importances = coef_or_feature_importances(LGBMRegressor(n_jobs = -1, random_state = my_seed, verbose = -1), folds)
lgbm_importances.iloc[:7]

,0,1,2,3,4,mean_coef_or_importance
shop_id,576,473,427,405,372,450.6
unique_items_in_shop,460,434,406,408,491,439.8
year,411,408,410,441,514,436.8
date_block_num,445,356,360,383,334,375.6
item_price,330,370,402,364,354,364.0
unique_items_in_cat,219,241,299,258,274,258.2
med_price_shop,164,159,215,298,235,214.2


In [116]:
# Mean feature importances across 5 folds of XGBoost (top 7 only)
xgb_importances = coef_or_feature_importances(XGBRegressor(n_jobs = -1, random_state = my_seed), folds)
xgb_importances.iloc[:7]

,0,1,2,3,4,mean_coef_or_importance
item_id,0.294833,0.220563,0.224927,0.395025,0.047091,0.236488
med_price_shop,0.040224,0.037634,0.094507,0.288312,0.309881,0.154112
unique_items_in_shop,0.216597,0.139927,0.091294,0.045087,0.077975,0.114176
shop_id,0.140726,0.157355,0.085660,0.056715,0.066453,0.101382
med_price_all_items,0.000000,0.120440,0.120051,0.001149,0.199962,0.088321
date_block_num,0.082870,0.078359,0.071445,0.043447,0.016105,0.058445
year,0.033873,0.058669,0.025845,0.070213,0.058631,0.049446


CatBoost takes a very long time to run my feature importances function, so I will not include it here.

Looking at the top 7 "most important features" output by different models, I will choose the following features: ```med_price_shop```, ```shop_id```, ```unique_items_in_shop```, ```item_price```.

_(If I have more time, I will develop a more systematic approach for feature selection, but this is what I'm doing at the moment)_

In [132]:
# Edit the features for each fold
fold_1_train_new = fold_1_train.loc[:, ['med_price_shop', 'shop_id', 'unique_items_in_shop', 'item_price', 'item_cnt_month']]
fold_1_test_new = fold_1_test.loc[:, ['med_price_shop', 'shop_id', 'unique_items_in_shop', 'item_price', 'item_cnt_month']]

fold_2_train_new = fold_2_train.loc[:, ['med_price_shop', 'shop_id', 'unique_items_in_shop', 'item_price', 'item_cnt_month']]
fold_2_test_new = fold_2_test.loc[:, ['med_price_shop', 'shop_id', 'unique_items_in_shop', 'item_price', 'item_cnt_month']]

fold_3_train_new = fold_3_train.loc[:, ['med_price_shop', 'shop_id', 'unique_items_in_shop', 'item_price', 'item_cnt_month']]
fold_3_test_new = fold_3_test.loc[:, ['med_price_shop', 'shop_id', 'unique_items_in_shop', 'item_price', 'item_cnt_month']]

fold_4_train_new = fold_4_train.loc[:, ['med_price_shop', 'shop_id', 'unique_items_in_shop', 'item_price', 'item_cnt_month']]
fold_4_test_new = fold_4_test.loc[:, ['med_price_shop', 'shop_id', 'unique_items_in_shop', 'item_price', 'item_cnt_month']]

fold_5_train_new = fold_5_train.loc[:, ['med_price_shop', 'shop_id', 'unique_items_in_shop', 'item_price', 'item_cnt_month']]
fold_5_test_new = fold_5_test.loc[:, ['med_price_shop', 'shop_id', 'unique_items_in_shop', 'item_price', 'item_cnt_month']]

In [138]:
# As an example
fold_5_train_new.head()

,med_price_shop,shop_id,unique_items_in_shop,item_price,item_cnt_month
0,196.0,0,2385,1322.0,10.0
1,196.0,0,2385,560.0,1.0
2,196.0,0,2385,806.0,4.0
3,196.0,0,2385,2231.0,5.0
4,196.0,0,2385,2381.0,1.0


In [133]:
# Split into features and targets again

# Fold 1
X_train_1_new = fold_1_train_new.drop('item_cnt_month', axis = 1).values
y_train_1_new = fold_1_train_new[['item_cnt_month']].values.ravel()
X_test_1_new = fold_1_test_new.drop('item_cnt_month', axis = 1).values
y_test_1_new = fold_1_test_new[['item_cnt_month']].values.ravel()

# Fold 2
X_train_2_new = fold_2_train_new.drop('item_cnt_month', axis = 1).values
y_train_2_new = fold_2_train_new[['item_cnt_month']].values.ravel()
X_test_2_new = fold_2_test_new.drop('item_cnt_month', axis = 1).values
y_test_2_new = fold_2_test_new[['item_cnt_month']].values.ravel()

# Fold 3
X_train_3_new = fold_3_train_new.drop('item_cnt_month', axis = 1).values
y_train_3_new = fold_3_train_new[['item_cnt_month']].values.ravel()
X_test_3_new = fold_3_test_new.drop('item_cnt_month', axis = 1).values
y_test_3_new = fold_3_test_new[['item_cnt_month']].values.ravel()

# Fold 4
X_train_4_new = fold_4_train_new.drop('item_cnt_month', axis = 1).values
y_train_4_new = fold_4_train_new[['item_cnt_month']].values.ravel()
X_test_4_new = fold_4_test_new.drop('item_cnt_month', axis = 1).values
y_test_4_new = fold_4_test_new[['item_cnt_month']].values.ravel()

# Fold 5
X_train_5_new = fold_5_train_new.drop('item_cnt_month', axis = 1).values
y_train_5_new = fold_5_train_new[['item_cnt_month']].values.ravel()
X_test_5_new = fold_5_test_new.drop('item_cnt_month', axis = 1).values
y_test_5_new = fold_5_test_new[['item_cnt_month']].values.ravel()

In [143]:
folds_new = [[X_train_1_new, X_test_1_new, y_train_1_new, y_test_1_new], [X_train_2_new, X_test_2_new, y_train_2_new, y_test_2_new],
             [X_train_3_new, X_test_3_new, y_train_3_new, y_test_3_new], [X_train_4_new, X_test_4_new, y_train_4_new, y_test_4_new],
             [X_train_5_new, X_test_5_new, y_train_5_new, y_test_5_new]]

# Remember to change the transformer accordingly with the new features
transformer_new = ColumnTransformer(
    transformers = [('target_enc', target_enc, [1]),
                    ('minmax_scaler', MinMaxScaler(), [0, 2, 3])],
    remainder = 'passthrough'
)

In [ ]:
model_list = [('Linear Regression', LinearRegression(n_jobs = -1), transformer_new, folds_new),
               ('Lasso', Lasso(random_state = my_seed), transformer_new, folds_new),
               ('Ridge', Ridge(random_state = my_seed), transformer_new, folds_new),
               ('Elastic Net', ElasticNet(random_state = my_seed), transformer_new, folds_new)]

overall_results = pd.concat([kfold_regression(model_name, model, transformer, folds) for model_name, model, transformer, folds in model_list], axis = 0, ignore_index = True)
overall_results.sort_values(['mean_rmse', 'std_rmse', 'execution_time'], inplace = True, ignore_index = True)
overall_results

,model_name,mean_rmse,std_rmse,execution_time
0,Elastic Net,1.995931,2.520029,1.539344
1,Lasso,1.995931,2.520029,1.644030
2,Ridge,2.003311,2.513683,1.508836
3,Linear Regression,2.003394,2.513621,1.840203


In [ ]:
# WARNING: THIS CODE BLOCK WILL TAKE A LONG TIME TO RUN (est. 30 mins)

model_list_2 = [('Decision Tree', DecisionTreeRegressor(random_state = my_seed), transformer_new, folds_new),
              ('Random Forest', RandomForestRegressor(n_jobs = -1, random_state = my_seed), transformer_new, folds_new),
              ('AdaBoost', AdaBoostRegressor(random_state = my_seed), transformer_new, folds_new),
              ('Gradient Boosting', GradientBoostingRegressor(random_state = my_seed), transformer_new, folds_new),
              ('Hist-based Gradient Boosting', HistGradientBoostingRegressor(random_state = my_seed), transformer_new, folds_new),
              ('XGBoost', XGBRegressor(n_jobs = -1, random_state = my_seed), transformer_new, folds_new),
              ('CatBoost', CatBoostRegressor(random_state = my_seed, verbose = False), transformer_new, folds_new),
              ('LightGBM', LGBMRegressor(n_jobs = -1, random_state = my_seed, verbose = -1), transformer_new, folds_new)]

overall_results = pd.concat([kfold_regression(model_name, model, transformer, folds) for model_name, model, transformer, folds in model_list_2], axis = 0, ignore_index = True)
overall_results.sort_values(['mean_rmse', 'std_rmse', 'execution_time'], inplace = True, ignore_index = True)
overall_results

,model_name,mean_rmse,std_rmse,execution_time
0,AdaBoost,1.116967,1.067410,218.973728
1,Gradient Boosting,1.244388,1.247079,579.035178
2,Random Forest,1.247933,1.686888,102.238504
3,LightGBM,1.938657,2.309566,22.255609
4,CatBoost,1.964565,1.797574,743.932278
5,Hist-based Gradient Boosting,1.966352,2.479180,27.655547
6,XGBoost,2.142089,2.394970,24.997484
7,Decision Tree,2.495480,2.137287,17.764580


### VII. Hyperparameter tuning

In [49]:
import optuna
from optuna.samplers import TPESampler

In [55]:
# Tuning for ElasticNet
def objective_elasticnet(trial):
    alpha = trial.suggest_float('alpha', 10**(-3), 10**2, log = True)
    l1_ratio = trial.suggest_float('l1_ratio', 0.05, 0.95, step = 0.05)
    tol = trial.suggest_float('tol', 10**(-5), 10**(-2), log = True)

    # Increase max_iter to 10000 for better results
    elasticnet_tuned = ElasticNet(random_state = my_seed, max_iter = 10000, alpha = alpha,
                                  l1_ratio = l1_ratio, tol = tol)

    try:
        score = kfold_regression('Elastic Net', elasticnet_tuned, folds, output = 'score')

        if trial.should_prune():
            raise optuna.TrialPruned()
    
        return score
    
    except Exception:
        pass

    # Use the try-except syntax to make sure the study does not end when a trial fails 

In [ ]:
# WARNING: THIS CODE BLOCK WILL TAKE A LONG TIME TO RUN (est. 2 hours)

# https://optuna.readthedocs.io/en/stable/reference/generated/optuna.pruners.MedianPruner.html
# Use MedianPruner as a form of early stopping
study1 = optuna.create_study(sampler = TPESampler(seed = my_seed), direction = 'minimize',
    pruner = optuna.pruners.MedianPruner(n_startup_trials = 3, n_warmup_steps = 1, interval_steps = 1, n_min_trials = 1))

# Not printing out Optuna updates periodically, including failed trials
optuna.logging.set_verbosity(optuna.logging.ERROR)
study1.optimize(objective_elasticnet, n_trials = 500)

# https://optuna.readthedocs.io/en/stable/reference/generated/optuna.trial.FrozenTrial.html
elasticnet_best_params = study1.best_trial.params

print(f'Trial number {study1.best_trial.number} has the best performance, with a score of {study1.best_trial.value}.')
print(f'The best parameters are: {study1.best_trial.params}')

[I 2025-08-07 10:43:25,233] A new study created in memory with name: no-name-75c7ddf2-97e8-4ce2-b8de-debafa0fdb06


Trial number 264 has the best performance, with a score of 1.995575197368847.
The best parameters are: {'alpha': 0.3668848655571989, 'l1_ratio': 0.35000000000000003, 'tol': 0.0027079448878869965}


In [59]:
# Tuning for LightGBM
# https://medium.com/@amitsinghrajput_92567/understanding-hyperparameters-in-decision-trees-xgboost-and-lightgbm-7b64cfed77f0

def objective_lgbm(trial):
    num_leaves = trial.suggest_int('num_leaves', 5, 100, step = 1)
    max_depth = trial.suggest_int('max_depth', 5, 100, step = 1)
    learning_rate = trial.suggest_float('learning_rate', 10**(-4), 10**(-1), log = True)
    min_split_gain = trial.suggest_float('min_split_gain', 0.05, 0.95, step = 0.05)
    min_child_weight = trial.suggest_float('min_child_weight', 10**(-4), 10**1, log = True)
    min_child_samples = trial.suggest_int('min_child_samples', 10, 10000, log = True)
    subsample = trial.suggest_float('subsample', 0.05, 0.95, step = 0.05)
    subsample_freq = trial.suggest_int('subsample_freq', 10, 1000, log = True)
    colsample_bytree = trial.suggest_float('colsample_bytree', 0.05, 0.95, step = 0.05)
    reg_alpha = trial.suggest_float('reg_alpha', 10**(-3), 10**1, log = True)
    reg_lambda = trial.suggest_float('reg_lambda', 10**(-3), 10**1, log = True)

    # Set n_estimators = 10000, random_state, verbose, and n_jobs to preferred value
    lgbm_tuned = LGBMRegressor(n_estimators = 10000, random_state = my_seed, n_jobs = -1, verbose = -1,
                               num_leaves = num_leaves, max_depth = max_depth, learning_rate = learning_rate,
                               min_split_gain = min_split_gain, min_child_weight = min_child_weight,
                               min_child_samples = min_child_samples, subsample = subsample,
                               subsample_freq = subsample_freq, colsample_bytree = colsample_bytree,
                               reg_alpha = reg_alpha, reg_lambda = reg_lambda)

    try:
        score = kfold_regression('LightGBM', lgbm_tuned, folds, output = 'score')

        if trial.should_prune():
            raise optuna.TrialPruned()
    
        return score
    
    except Exception:
        pass

    # Use the try-except syntax to make sure the study does not end when a trial fails

In [62]:
# WARNING: THIS CODE BLOCK WILL TAKE A LONG TIME TO RUN

study = optuna.create_study(sampler = TPESampler(seed = my_seed), direction = 'minimize',
    pruner = optuna.pruners.MedianPruner(n_startup_trials = 3, n_warmup_steps = 1, interval_steps = 1, n_min_trials = 1))

# optuna.logging.set_verbosity(optuna.logging.ERROR) # can uncomment this command to avoid excessive logging by Optuna
study.optimize(objective_lgbm, n_trials = 10)

lgbm_best_params = study.best_trial.params

print(f'Trial number {study.best_trial.number} has the best performance, with a score of {study.best_trial.value}.')
print(f'The best parameters are: {study.best_trial.params}')

Trial number 0 has the best performance, with a score of 1.5233667675060847.
The best parameters are: {'num_leaves': 31, 'max_depth': 64, 'learning_rate': 0.023251443377156827, 'min_split_gain': 0.6000000000000001, 'min_child_weight': 0.053160726817981174, 'min_child_samples': 64, 'subsample': 0.8, 'subsample_freq': 189, 'colsample_bytree': 0.95, 'reg_alpha': 0.31549086474120447, 'reg_lambda': 0.08921894140014044}


### VIII. Prepare data for final submission

In [48]:
# Declare final model for submission
from sklearn.ensemble import StackingRegressor

#final_model = StackingRegressor(estimators = [('elasticnet', ElasticNet(random_state = my_seed, max_iter = 10000, **elasticnet_best_params)),
                                              #('lightgbm', LGBMRegressor(n_estimators = 10000, random_state = my_seed, n_jobs = -1, verbose = -1, **lgbm_best_params))],
                                #n_jobs = -1)

In [49]:
test_shop_items = test[['shop_id', 'item_id']].copy()
test_shop_items.head() # this is the test data stripped of its ID

,shop_id,item_id
0,5,5037
1,5,5320
2,5,5233
3,5,5232
4,5,5268


In [50]:
# Set constant data for some columns
test_features = test_shop_items.copy()

test_features['year'] = 2013
test_features['date_block_num'] = 34 # number of months elapsed since Jan 2013

test_features['spring'] = 0; test_features['summer'] = 0; test_features['fall'] = 1 # November is in the fall season
test_features['weekends_in_month'] = 9
test_features['holidays_in_month'] = 1 # Russia's Unity Day in 4 November

# For the month_sin and month_cos trigonometric encoding, replace the month number with 11
test_features['month_sin'] = np.sin(11 / 12 * 2 * np.pi)
test_features['month_cos'] = np.cos(11 / 12 * 2 * np.pi)

test_features.head()

,shop_id,item_id,year,date_block_num,spring,summer,fall,weekends_in_month,holidays_in_month,month_sin,month_cos
0,5,5037,2013,34,0,0,1,9,1,-0.5,0.866025
1,5,5320,2013,34,0,0,1,9,1,-0.5,0.866025
2,5,5233,2013,34,0,0,1,9,1,-0.5,0.866025
3,5,5232,2013,34,0,0,1,9,1,-0.5,0.866025
4,5,5268,2013,34,0,0,1,9,1,-0.5,0.866025


In [51]:
# Retrieve item category and item price (each item only belongs to 1 category and has 1 price)
item_info = final_features[['shop_id', 'item_id', 'item_category_id', 'item_price']].copy()

# Keep only the last occurence of (shop_id, item_id) pair (showing the latest price)
item_info.drop_duplicates(subset = ['shop_id', 'item_id'], keep = 'last', inplace = True, ignore_index = True)
item_info.head()

,shop_id,item_id,item_category_id,item_price
0,0,5572,2,1322.0
1,0,5575,2,806.0
2,0,5576,2,2231.0
3,0,5612,2,3623.0
4,0,5623,2,294.0


In [52]:
# Merge item information to the test set
item_info = test_shop_items.merge(item_info, on = ['shop_id', 'item_id'], how = 'left')
item_info.drop_duplicates(inplace = True, ignore_index = True)
item_info.head()

,shop_id,item_id,item_category_id,item_price
0,5,5037,19.0,749.5
1,5,5320,NaN,NaN
2,5,5233,19.0,1199.0
3,5,5232,23.0,599.0
4,5,5268,NaN,NaN


In [53]:
item_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 214200 entries, 0 to 214199
Data columns (total 4 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   shop_id           214200 non-null  int64  
 1   item_id           214200 non-null  int64  
 2   item_category_id  104268 non-null  float64
 3   item_price        104268 non-null  float64
dtypes: float64(2), int64(2)
memory usage: 6.5 MB


In [54]:
# Create a new item category ID that denotes "unknown"
new_cat_id = item_categories['item_category_id'].max() + 1 # new identifier for unknown category
print(new_cat_id)

84


In [55]:
# item_info['item_category_id']
item_info['item_category_id'] = item_info['item_category_id'].fillna(new_cat_id)
item_info.head()

,shop_id,item_id,item_category_id,item_price
0,5,5037,19.0,749.5
1,5,5320,84.0,NaN
2,5,5233,19.0,1199.0
3,5,5232,23.0,599.0
4,5,5268,84.0,NaN


In [56]:
from sklearn.impute import SimpleImputer
imp = SimpleImputer(strategy = 'median')

imputed_data = imp.fit_transform(item_info) # outputs a NumPy array
new_item_info = pd.DataFrame(imputed_data, columns = item_info.columns) # convert into a DataFrame
new_item_info.head()

,shop_id,item_id,item_category_id,item_price
0,5.0,5037.0,19.0,749.5
1,5.0,5320.0,84.0,449.0
2,5.0,5233.0,19.0,1199.0
3,5.0,5232.0,23.0,599.0
4,5.0,5268.0,84.0,449.0


In [57]:
# Create new features based on imputed data (for shops)
new_item_info_shop = new_item_info[['shop_id', 'item_id', 'item_price']].copy()
new_item_info_shop = new_item_info_shop.groupby('shop_id', as_index = False).agg({'item_price': 'median', 'item_id': 'nunique'})
new_item_info_shop.rename(columns = {'item_price': 'med_price_shop', 'item_id': 'unique_items_in_shop'}, inplace = True)
new_item_info_shop.head()

,shop_id,med_price_shop,unique_items_in_shop
0,2.0,449.0,5100
1,3.0,449.0,5100
2,4.0,449.0,5100
3,5.0,449.0,5100
4,6.0,449.0,5100


In [58]:
# Create new features based on imputed data (for categories)
new_item_info_cat = new_item_info[['item_category_id', 'item_id', 'item_price']].copy()
new_item_info_cat = new_item_info_cat.groupby('item_category_id', as_index = False).agg({'item_price': 'median', 'item_id': 'nunique'})
new_item_info_cat.rename(columns = {'item_price': 'med_price_cat', 'item_id': 'unique_items_in_cat'}, inplace = True)
new_item_info_cat.head()

,item_category_id,med_price_cat,unique_items_in_cat
0,2.0,2990.0,10
1,3.0,1390.0,27
2,5.0,599.0,4
3,6.0,2290.0,9
4,7.0,2489.5,23


In [59]:
# Get median price of all items (including imputed) sold in November 2015 - test data 
new_item_info['med_price_all_items'] = new_item_info['item_price'].median()
new_item_info.head()

,shop_id,item_id,item_category_id,item_price,med_price_all_items
0,5.0,5037.0,19.0,749.5,449.0
1,5.0,5320.0,84.0,449.0,449.0
2,5.0,5233.0,19.0,1199.0,449.0
3,5.0,5232.0,23.0,599.0,449.0
4,5.0,5268.0,84.0,449.0,449.0


In [60]:
# Merge into the existing new_item_info DataFrame
item_info = new_item_info.merge(new_item_info_shop, how = 'left', on = 'shop_id')\
    .merge(new_item_info_cat, how = 'left', on = 'item_category_id')

item_info.head()

,shop_id,item_id,item_category_id,item_price,med_price_all_items,med_price_shop,unique_items_in_shop,med_price_cat,unique_items_in_cat
0,5.0,5037.0,19.0,749.5,449.0,449.0,5100,1199.0,137
1,5.0,5320.0,84.0,449.0,449.0,449.0,5100,449.0,5100
2,5.0,5233.0,19.0,1199.0,449.0,449.0,5100,1199.0,137
3,5.0,5232.0,23.0,599.0,449.0,449.0,5100,1199.0,143
4,5.0,5268.0,84.0,449.0,449.0,449.0,5100,449.0,5100


In [61]:
# Now merge to create the full test set
test_features = test_features.merge(item_info, how = 'left', on = ['shop_id', 'item_id'])
test_features.head()

,shop_id,item_id,year,date_block_num,spring,summer,fall,weekends_in_month,holidays_in_month,month_sin,month_cos,item_category_id,item_price,med_price_all_items,med_price_shop,unique_items_in_shop,med_price_cat,unique_items_in_cat
0,5,5037,2013,34,0,0,1,9,1,-0.5,0.866025,19.0,749.5,449.0,449.0,5100,1199.0,137
1,5,5320,2013,34,0,0,1,9,1,-0.5,0.866025,84.0,449.0,449.0,449.0,5100,449.0,5100
2,5,5233,2013,34,0,0,1,9,1,-0.5,0.866025,19.0,1199.0,449.0,449.0,5100,1199.0,137
3,5,5232,2013,34,0,0,1,9,1,-0.5,0.866025,23.0,599.0,449.0,449.0,5100,1199.0,143
4,5,5268,2013,34,0,0,1,9,1,-0.5,0.866025,84.0,449.0,449.0,449.0,5100,449.0,5100


In [62]:
# Create submission training and test data
submission_train = final_features.copy()

# After checking data types (truncated), I need to change data type of the weekends_in_month and holidays_in_month columns from object to int
submission_train['weekends_in_month'] = submission_train['weekends_in_month'].astype('int')
submission_train['holidays_in_month'] = submission_train['weekends_in_month'].astype('int')

submission_train.head()

,year,date_block_num,shop_id,item_category_id,item_id,item_price,item_cnt_month,med_price_shop,unique_items_in_shop,med_price_cat,unique_items_in_cat,med_price_all_items,weekends_in_month,holidays_in_month,spring,summer,fall,month_sin,month_cos
0,2013,0,0,2,5572,1322.0,10.0,196.0,2385,2390.0,37,299.0,8,8,0,0,0,0.5,0.866025
1,2013,0,0,2,5573,560.0,1.0,196.0,2385,2390.0,37,299.0,8,8,0,0,0,0.5,0.866025
2,2013,0,0,2,5575,806.0,4.0,196.0,2385,2390.0,37,299.0,8,8,0,0,0,0.5,0.866025
3,2013,0,0,2,5576,2231.0,5.0,196.0,2385,2390.0,37,299.0,8,8,0,0,0,0.5,0.866025
4,2013,0,0,2,5609,2381.0,1.0,196.0,2385,2390.0,37,299.0,8,8,0,0,0,0.5,0.866025


In [63]:
# Reorder the columns to align with the training data
submission_test = test_features.loc[:, final_features.drop('item_cnt_month', axis = 1).columns.tolist()]
submission_test.head()

,year,date_block_num,shop_id,item_category_id,item_id,item_price,med_price_shop,unique_items_in_shop,med_price_cat,unique_items_in_cat,med_price_all_items,weekends_in_month,holidays_in_month,spring,summer,fall,month_sin,month_cos
0,2013,34,5,19.0,5037,749.5,449.0,5100,1199.0,137,449.0,9,1,0,0,1,-0.5,0.866025
1,2013,34,5,84.0,5320,449.0,449.0,5100,449.0,5100,449.0,9,1,0,0,1,-0.5,0.866025
2,2013,34,5,19.0,5233,1199.0,449.0,5100,1199.0,137,449.0,9,1,0,0,1,-0.5,0.866025
3,2013,34,5,23.0,5232,599.0,449.0,5100,1199.0,143,449.0,9,1,0,0,1,-0.5,0.866025
4,2013,34,5,84.0,5268,449.0,449.0,5100,449.0,5100,449.0,9,1,0,0,1,-0.5,0.866025


Do the train-test split and implementing a pipeline process to submit.

In [80]:
# Full pipeline
X_train_submission = submission_train.drop('item_cnt_month', axis = 1).values
y_train_submission = submission_train[['item_cnt_month']].values.ravel()
X_test_submission = submission_test.values

full_pipe = Pipeline([('target_enc', transformer), ('regression', final_model)])

full_pipe.fit(X_train_submission, y_train_submission)

submission_test['item_cnt_month_unadjusted'] = full_pipe.predict(X_test_submission)

# Cap the predicted values in the [0, 20] range
submission_test['item_cnt_month'] = submission_test['item_cnt_month_unadjusted'].case_when\
    ([(submission_test['item_cnt_month_unadjusted'] < 0, 0),
      (submission_test['item_cnt_month_unadjusted'] > 20, 20),
      (np.logical_and(submission_test['item_cnt_month_unadjusted'] >= 0, submission_test['item_cnt_month_unadjusted'] <= 20),
       submission_test['item_cnt_month_unadjusted'])])

submission_test.head()

,year,date_block_num,shop_id,item_category_id,item_id,item_price,med_price_shop,unique_items_in_shop,med_price_cat,unique_items_in_cat,med_price_all_items,weekends_in_month,holidays_in_month,spring,summer,fall,month_sin,month_cos,item_cnt_month_unadjusted,item_cnt_month
0,2013,34,5,19.0,5037,749.5,449.0,5100,1199.0,137,449.0,9,1,0,0,1,-0.5,0.866025,1.118458,1.118458
1,2013,34,5,84.0,5320,449.0,449.0,5100,449.0,5100,449.0,9,1,0,0,1,-0.5,0.866025,1.414916,1.414916
2,2013,34,5,19.0,5233,1199.0,449.0,5100,1199.0,137,449.0,9,1,0,0,1,-0.5,0.866025,1.593280,1.593280
3,2013,34,5,23.0,5232,599.0,449.0,5100,1199.0,143,449.0,9,1,0,0,1,-0.5,0.866025,2.004942,2.004942
4,2013,34,5,84.0,5268,449.0,449.0,5100,449.0,5100,449.0,9,1,0,0,1,-0.5,0.866025,1.414916,1.414916


In [81]:
# Extract only the necessary columns
submission_without_id = submission_test[['shop_id', 'item_id', 'item_cnt_month']].copy()

# Merge the ID column, then remove the shop_id and item_id column to fit with sample submission
submission = submission_without_id.merge(test, on = ['shop_id', 'item_id'])
submission = submission.loc[:, ['ID', 'item_cnt_month']]
print(submission.shape)
submission.head()

(214200, 2)


,ID,item_cnt_month
0,0,1.118458
1,1,1.414916
2,2,1.593280
3,3,2.004942
4,4,1.414916


In [82]:
# Export submission to .csv file and send to Kaggle
submission.to_csv('thai_an_le_submission.csv', index = False)